src/ImageSharp/Formats/Png/PngDecoderCore.cs

```diff
@@ -585,22 +714,25 @@ private void ReadScanlines<TPixel>(PngChunk chunk, ImageFrame<TPixel> image, Png
     private void DecodePixelData<TPixel>(DeflateStream compressedStream, ImageFrame<TPixel> image, PngMetadata pngMetadata, CancellationToken cancellationToken)
         where TPixel : unmanaged, IPixel<TPixel>
     {
-        while (this.currentRow < this.header.Height)
+        int currentRow = Adam7.FirstRow[0];
+        int currentRowBytesRead = 0;
+        int height = image.Metadata.TryGetAPngFrameMetadata(out APngFrameMetadata? frameMetadata) ? frameMetadata.Height : this.header.Height;
+        while (currentRow < height)
```

Comment:
All frames within a decoded image must be the same size. The decoded data should be offset against the frame. See the gif decoder.